File: java-start/app/src/main/java/com/vonage/tutorial/voice/view/login/LoginViewModel.java
package com.vonage.tutorial.voice.view.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.navigation.NavDirections;
import com.nexmo.client.NexmoClient;
import com.nexmo.client.request_listener.NexmoConnectionListener.ConnectionStatus;
import com.vonage.tutorial.voice.User;
import com.vonage.tutorial.voice.util.NavManager;
import com.vonage.tutorial.voice.util.Tutorial;

public class LoginViewModel extends ViewModel {

    private NexmoClient client = null; // ToDo: Initialize

    NavManager navManager = NavManager.getInstance();

    private MutableLiveData<ConnectionStatus> connectionStatusMutableLiveData = new MutableLiveData<ConnectionStatus>();
    public LiveData<ConnectionStatus> _connectionStatusLiveData = connectionStatusMutableLiveData;

    private User user;


    public LoginViewModel() {
        Tutorial.todo("Add status connection listener");
    }

    private void navigate() {
        String userName = user.getName();

        if(userName == null) {
            throw new RuntimeException("user name is null");
        }

        NavDirections navDirections = LoginFragmentDirections.actionLoginFragmentToVoiceFragment(userName);
        navManager.navigate(navDirections);
    }

    void onLoginUser(User user) {
        Tutorial.todo("Login user");
    }
}

File: kotlin-start/app/src/main/java/com/vonage/tutorial/voice/view/login/LoginViewModel.kt
package com.vonage.tutorial.voice.view.login

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.nexmo.client.NexmoClient
import com.nexmo.client.request_listener.NexmoConnectionListener.ConnectionStatus
import com.vonage.tutorial.voice.User
import com.vonage.tutorial.voice.extension.asLiveData
import com.vonage.tutorial.voice.util.NavManager

class LoginViewModel : ViewModel() {

    private val navManager = NavManager

    private val _connectionStatus = MutableLiveData<ConnectionStatus>()
    val connectionStatus = _connectionStatus.asLiveData()

    private var user: User? = null

    private val client: NexmoClient = TODO("Retrieve NexmoClient instance")

    init {
        client.setConnectionListener { newConnectionStatus, _ ->

            if (newConnectionStatus == ConnectionStatus.CONNECTED) {
                navigateToVoiceFragment()
                return@setConnectionListener
            }

            _connectionStatus.postValue(newConnectionStatus)
        }
    }

    private fun navigateToVoiceFragment() {
        val userName = checkNotNull(user?.name) { "user is null" }
        val navDirections =
            LoginFragmentDirections.actionLoginFragmentToVoiceFragment(
                userName
            )
        navManager.navigate(navDirections)
    }

    fun onLoginUser(user: User) {
        // TODO: Login user
    }
}
